File: src/main/java/org/jukeboxmc/block/behavior/BlockSapling.java
package org.jukeboxmc.block.behavior;

import com.nukkitx.nbt.NbtMap;
import org.jetbrains.annotations.NotNull;
import org.jukeboxmc.block.Block;
import org.jukeboxmc.block.data.SaplingType;
import org.jukeboxmc.item.Item;
import org.jukeboxmc.item.ItemType;
import org.jukeboxmc.item.behavior.ItemSapling;
import org.jukeboxmc.util.Identifier;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class BlockSapling extends Block {

    public BlockSapling( Identifier identifier ) {
        super( identifier );
    }

    public BlockSapling( Identifier identifier, NbtMap blockStates ) {
        super( identifier, blockStates );
    }

    @Override
    public Item toItem() {
        return Item.<ItemSapling>create( ItemType.SAPLING ).setSaplingType( this.getSaplingType() );
    }

    public BlockSapling setSaplingType(@NotNull SaplingType saplingType ) {
        return this.setState( "sapling_type", saplingType.name().toLowerCase() );
    }

    public @NotNull SaplingType getSaplingType() {
        return this.stateExists( "sapling_type" ) ? SaplingType.valueOf( this.getStringState( "sapling_type" ) ) : SaplingType.OAK;
    }

    public void setAge( boolean value ) {
        this.setState( "age_bit", value ? (byte) 1 : (byte) 0 );
    }

    public boolean hasAge() {
        return this.stateExists( "age_bit" ) && this.getByteState( "age_bit" ) == 1;
    }
}

File: src/main/java/org/jukeboxmc/block/behavior/BlockSapling.kt
package org.jukeboxmc.block.behavior

import com.nukkitx.nbt.NbtMap
import org.jukeboxmc.block.Block
import org.jukeboxmc.block.data.SaplingType
import org.jukeboxmc.item.Item
import org.jukeboxmc.item.ItemType
import org.jukeboxmc.item.behavior.ItemSapling
import org.jukeboxmc.util.Identifier
import java.util.Locale

/**
 * @author LucGamesYT
 * @version 1.0
 */
class BlockSapling : Block {
    constructor(identifier: Identifier?) : super(identifier)
    constructor(identifier: Identifier?, blockStates: NbtMap?) : super(identifier, blockStates)

    override fun toItem(): Item {
        return Item.create<ItemSapling>(ItemType.SAPLING).setSaplingType(
            saplingType,
        )
    }

    fun setSaplingType(saplingType: SaplingType): BlockSapling {
        return setState("sapling_type", saplingType.name.lowercase(Locale.getDefault()))
    }

    val saplingType: SaplingType
        get() = if (stateExists("sapling_type")) SaplingType.valueOf(getStringState("sapling_type")) else SaplingType.OAK

    fun setAge(value: Boolean) {
        setState<Block>("age_bit", if (value) 1.toByte() else 0.toByte())
    }

    fun hasAge(): Boolean {
        return stateExists("age_bit") && getByteState("age_bit").toInt() == 1
    }
}
